labs/a8.md

```diff
@@ -0,0 +1,736 @@
+understanding your threat model. No system is guaranteed to be secure or able 
+to withstand all attacks, nor is this even possible in the face of extreme 
+adversaries. However, you can (and should) take precautions against the threats 
+you are likely to face. Balancing the need for authorized users to get access 
+to the system while keeping unauthorized users out is very easy to get wrong. 
+Fortunately, smart people have distilled the principles of security down to a 
+few axioms, covered very well in the first [lecture note][cs161_1] of CS 161 
+(credit to Prof. David Wagner). Please read the note prior to continuing.
+
+[cs161_1]: http://www.icir.org/vern/cs161-sp17/notes/Principles.1.19.pdf
+
+## Security Building Blocks
+
+### Encryption
+
+Hiding information from those who are not authorized to view it is a critical 
+element of computer security. How can you make sure this property can be 
+maintained in the event of having to transport this information in the open? By 
+encrypting it. 
+
+Encryption involves taking the information you want to hide, called the 
+_plaintext_, and transforming it by means of a mathematical function into a 
+format that is indecipherable without a key, called the _ciphertext_. There are 
+many different types of encryption algorithms, with different types of 
+encryption keys, encryption speeds, security, and usefulness. The most 
+important quality of encryption algorithms is the fact that they are one-way: 
+it is easy to compute the ciphertext from the plaintext (encrypt) but extremely 
+difficult to compute the plaintext from the ciphertext (decrypt) without the 
+secret key.
+
+There are two primary types of encryption: **symmetric key cryptography**, and 
+asymmetric-key or **public-key cryptography**. The invention of public key 
+cryptography was critical to many of the security features we take for granted 
+today.
+
+For the purpose of this lab, we will be using several tools for performing 
+encryption operations, among them **GnuPG**, a free implementation of the 
+[OpenPGP](https://openpgp.org/) standard, and 
+[**OpenSSL**](https://openssl.com), a library for implementing TLS, or 
+Transport Layer Security.
+
+##### Symmetric Key Cryptography
+
+Symmetric-key cryptography is named for the fact that a single key both 
+encrypts and decrypts a particular plaintext to ciphertext and vice-versa. The 
+most widely used method of symmetric crypto is AES, or _Advanced Encryption 
+Standard_, which is certified and trusted by the US Government to encrypt 
+information critical to national security. An easy way to visualize the basics 
+of symmetric key crypto is by the XOR function: `a XOR b` is true if and only 
+if a and b are different. Suppose our plaintext is the bitstring  
+`100110100101`, and we want to apply a bitwise XOR with the bitstring 
+`010010100010`, which is our key. The result of this operation is 
+`110100000111`. If we XOR the resulting ciphertext with the key again, we get 
```

Comment:
I think this is better explained through a diagram (since then the digits could be lined up). Wikipedia probably has one.

Reply:
done